akka-actor/src/main/scala/akka/util/ByteString.scala

```diff
@@ -364,21 +364,57 @@ object ByteString {
   }
 
   /**
+   * This class helps to optimise the access time for linear access such as looping all items of [[ByteStrings]].
+   *
+   * @param lastIndex the index of the last access
+   * @param arrayIndex the array index of the last access
+   * @param seen the amount of elements being seen, excluding the elements in current array
+   */
+  private final case class LinearScanHint(lastIndex: Int, arrayIndex: Int, seen: Int)
```

Comment:
Yep, as @2beaucoup said.
It does not use more space and alg could be updated like so.
We'd want this to be a view still though (return wrapper with the counter).

If you don't believe object size guessing you can use jol:

```
case class LinearScanHintInts(lastIndex: Int, arrayIndex: Int, seen: Int)

[info] ***** 64-bit VM: **********************************************************
[info] akka.stream.LinearScanHintInts object internals:
[info]  OFFSET  SIZE  TYPE DESCRIPTION                    VALUE
[info]       0    16       (object header)                N/A
[info]      16     4   int LinearScanHintInts.lastIndex   N/A
[info]      20     4   int LinearScanHintInts.arrayIndex  N/A
[info]      24     4   int LinearScanHintInts.seen        N/A
[info]      28     4       (loss due to the next object alignment)
[info] Instance size: 32 bytes
[info] Space losses: 0 bytes internal + 4 bytes external = 4 bytes total
[info]
[info] ***** 64-bit VM, compressed references enabled: ***************************
[info] akka.stream.LinearScanHintInts object internals:
[info]  OFFSET  SIZE  TYPE DESCRIPTION                    VALUE
[info]       0    12       (object header)                N/A
[info]      12     4   int LinearScanHintInts.lastIndex   N/A
[info]      16     4   int LinearScanHintInts.arrayIndex  N/A
[info]      20     4   int LinearScanHintInts.seen        N/A
[info] Instance size: 24 bytes
[info] Space losses: 0 bytes internal + 0 bytes external = 0 bytes total
[info]
[info] ***** 64-bit VM, compressed references enabled, 16-byte align: ************
[info] akka.stream.LinearScanHintInts object internals:
[info]  OFFSET  SIZE  TYPE DESCRIPTION                    VALUE
[info]       0    12       (object header)                N/A
[info]      12     4   int LinearScanHintInts.lastIndex   N/A
[info]      16     4   int LinearScanHintInts.arrayIndex  N/A
[info]      20     4   int LinearScanHintInts.seen        N/A
[info]      24     8       (loss due to the next object alignment)
```

```
case class LinearScanHintBs(current: ByteString, offset: Int)

[info] ***** 64-bit VM: **********************************************************
[info] akka.stream.LinearScanHintBs object internals:
[info]  OFFSET  SIZE       TYPE DESCRIPTION                    VALUE
[info]       0    16            (object header)                N/A
[info]      16     8 ByteString LinearScanHintBs.current       N/A
[info]      24     4        int LinearScanHintBs.offset        N/A
[info]      28     4            (loss due to the next object alignment)
[info] Instance size: 32 bytes
[info] Space losses: 0 bytes internal + 4 bytes external = 4 bytes total
[info]
[info] ***** 64-bit VM, compressed references enabled: ***************************
[info] akka.stream.LinearScanHintBs object internals:
[info]  OFFSET  SIZE       TYPE DESCRIPTION                    VALUE
[info]       0    12            (object header)                N/A
[info]      12     4 ByteString LinearScanHintBs.current       N/A
[info]      16     4        int LinearScanHintBs.offset        N/A
[info]      20     4            (loss due to the next object alignment)
[info] Instance size: 24 bytes
[info] Space losses: 0 bytes internal + 4 bytes external = 4 bytes total
```